zcash_primitives/src/memo.rs

```diff
@@ -0,0 +1,411 @@
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    InvalidUtf8(std::str::Utf8Error),
+    TooLong(usize),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidUtf8(e) => write!(f, "Invalid UTF-8: {}", e),
+            Error::TooLong(n) => write!(f, "Memo length {} is larger than maximum of 512", n),
+        }
+    }
+}
+
+impl error::Error for Error {}
+
+/// The unencrypted memo bytes received alongside a shielded note in a Zcash transaction.
+#[derive(Clone)]
+pub struct MemoBytes(pub(crate) Box<[u8; 512]>);
+
+impl fmt::Debug for MemoBytes {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "MemoBytes(")?;
+        fmt_colon_delimited_hex(f, &self.0[..])?;
+        write!(f, ")")
+    }
+}
+
+impl Default for MemoBytes {
+    fn default() -> Self {
+        let mut bytes = [0u8; 512];
+        bytes[0] = 0xF6;
+        MemoBytes(Box::new(bytes))
+    }
+}
+
+impl PartialEq for MemoBytes {
+    fn eq(&self, rhs: &MemoBytes) -> bool {
+        self.0[..] == rhs.0[..]
+    }
+}
+
+impl Eq for MemoBytes {}
+
+impl PartialOrd for MemoBytes {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for MemoBytes {
+    fn cmp(&self, rhs: &Self) -> Ordering {
+        self.0[..].cmp(&rhs.0[..])
+    }
+}
+
+impl MemoBytes {
+    /// Creates a `MemoBytes` from a slice.
+    ///
+    /// Returns an error if the provided slice is longer than 512 bytes. Slices shorter
+    /// than 512 bytes are padded with null bytes.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        if bytes.len() > 512 {
+            return Err(Error::TooLong(bytes.len()));
+        }
+
+        let mut memo = [0u8; 512];
+        memo[..bytes.len()].copy_from_slice(bytes);
+        Ok(MemoBytes(Box::new(memo)))
+    }
+
+    /// Returns the raw byte array containing the memo bytes, including null padding.
+    pub fn as_array(&self) -> &[u8; 512] {
+        &self.0
+    }
+
+    /// Returns a slice of the raw bytes, excluding null padding.
+    pub fn as_slice(&self) -> &[u8] {
+        let first_null = self
+            .0
+            .iter()
+            .enumerate()
+            .rev()
+            .find(|(_, &b)| b != 0)
+            .map(|(i, _)| i + 1)
+            .unwrap_or_default();
+
+        &self.0[..first_null]
+    }
+}
+
+/// Type-safe wrapper around String to enforce memo length requirements.
+#[derive(Clone, PartialEq)]
+pub struct TextMemo(String);
+
+impl From<TextMemo> for String {
+    fn from(memo: TextMemo) -> String {
+        memo.0
+    }
+}
+
+impl Deref for TextMemo {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &str {
+        self.0.deref()
+    }
+}
+
+/// An unencrypted memo received alongside a shielded note in a Zcash transaction.
+#[derive(Clone)]
+pub enum Memo {
```

Comment:
Should we actually move this to zcash_client_backend? There's nothing else in the zcash_primitives crate that depends upon this, and it really does seem entirely client-related.

Reply:
Nothing directly, but there may be other downstream users relying on it (who are otherwise using the note decryption APIs in `zcash_primitives` directly). Let's defer this to a later PR in any case.